src/main/java/io/openio/sds/storage/rawx/RawxClient.java

```diff
@@ -218,149 +204,149 @@ public void deleteChunks(List<ChunkInfo> l) {
-    }
+	private ObjectInfo uploadPosition(final ObjectInfo oinf, final int pos, final Long size,
+									  InputStream data, final String reqId) {
+		List<ChunkInfo> cil = oinf.sortedChunks().get(pos);
+		final List<FeedableInputStream> gens = size == 0 ? null : feedableBodies(cil.size(), size);
+		List<Future<UploadResult>> futures = new ArrayList<Future<UploadResult>>();
+		for (int i = 0; i < cil.size(); i++) {
+			final ChunkInfo ci = cil.get(i);
+			final FeedableInputStream in = null == gens ? null : gens.get(i);
+
+			Callable<UploadResult> uploader = new Callable<UploadResult>() {
+
+				@Override
+				public UploadResult call() {
+					UploadResult result = new UploadResult(ci);
+					try {
+						RequestBuilder builder = http
+								.put(ci.url())
+								.header(CHUNK_META_CONTAINER_ID, oinf.url().cid())
+								.header(CHUNK_META_CONTENT_ID, oinf.oid())
+								.header(CHUNK_META_CONTENT_VERSION, String.valueOf(oinf.version()))
+								.header(CHUNK_META_CONTENT_POLICY, oinf.policy())
+								.header(CHUNK_META_CONTENT_MIME_TYPE, oinf.mtype())
+								.header(CHUNK_META_CONTENT_CHUNK_METHOD, oinf.chunkMethod())
+								.header(CHUNK_META_CONTENT_CHUNKSNB,
+										String.valueOf(oinf.nbchunks()))
+								.header(CHUNK_META_CONTENT_SIZE, String.valueOf(oinf.size()))
+								.header(CHUNK_META_CONTENT_PATH, oinf.url().object())
+								.header(CHUNK_META_CHUNK_ID, ci.id())
+								.header(CHUNK_META_CHUNK_POS, ci.pos().toString())
+								.header(CHUNK_META_FULL_PATH, oinf.url().toFullPath())
+								.header(CHUNK_META_OIO_VERSION, "4")
+								.header(OIO_REQUEST_ID_HEADER, reqId).verifier(RAWX_VERIFIER);
+						if (null == gens)
+							builder.body("");
+						else
+							builder.body(in, size);
+						ci.size(size);
+						ci.hash(builder.execute().close(false).header(CHUNK_META_CHUNK_HASH));
+					} catch (OioException e) {
+						result.exception(e);
+					}
+					return result;
+				}
+			};
+			int retry = 0;
+			try {
+				while (true) {
+					try {
+						Future<UploadResult> upload = executors.submit(uploader);
+						futures.add(upload);
+						break;
+					} catch (RejectedExecutionException ree) {
+						if (retry < 5) {
+							int delay = 1 << retry;
+							logger.warn("Failed to start rawx upload, retry in " + delay + "s",
+									ree);
+							try {
+								Thread.sleep(delay * 1000);
+							} catch (InterruptedException e) {
+								throw new OioException("Failed to retry rawx upload", e);
+							}
+						} else {
+							throw new OioException("Failed to schedule rawx upload", ree);
+						}
+						retry++;
+					}
+				}
+			} catch (RuntimeException e) {
+				try {
+					in.close();
+				} catch (IOException e1) {
+					logger.warn(e1);
+				}
+				throw e;
+			}
+		}
+		consume(data, size, gens, futures);
+
+		try {
+			ArrayList<UploadResult> successes = new ArrayList<UploadResult>();
+			for (Future<UploadResult> f : futures) {
+				UploadResult result = f.get();
```

Comment:
Why did you get rid of the timeout?

Reply:
I don't see why such a timeout should be used to limit the time of a single chunk complete upload.

Reply:
This was a safeguard against stuck upload tasks. We can't configure write timeout on Java sockets.

Reply:
I understand but this safeguard is more harmful, it should not apply to the whole upload operation.